src/envoy/http/authn/authn_utils_test.cc

```diff
@@ -0,0 +1,164 @@
+namespace Http {
+namespace Istio {
+namespace AuthN {
+namespace {
+
+const std::string kSecIstioAuthUserInfoHeaderKey = "sec-istio-auth-userinfo";
+const std::string kSecIstioAuthUserinfoHeaderValue =
+    R"(
+     {
+       "iss": "628645741881-noabiu23f5a8m8ovd8ucv698lj78vv0l@developer.gserviceaccount.com",
+       "sub": "628645741881-noabiu23f5a8m8ovd8ucv698lj78vv0l@developer.gserviceaccount.com",
+       "aud": "bookstore-esp-echo.cloudendpointsapis.com",
+       "iat": 1512754205,
+       "exp": 5112754205
+     }
+   )";
+const std::string kSecIstioAuthUserInfoHeaderWithNoAudValue =
+    R"(
+       {
+         "iss": "628645741881-noabiu23f5a8m8ovd8ucv698lj78vv0l@developer.gserviceaccount.com",
+         "sub": "628645741881-noabiu23f5a8m8ovd8ucv698lj78vv0l@developer.gserviceaccount.com",
+         "iat": 1512754205,
+         "exp": 5112754205
+       }
+     )";
+const std::string kSecIstioAuthUserInfoHeaderWithTwoAudValue =
+    R"(
+       {
+         "iss": "628645741881-noabiu23f5a8m8ovd8ucv698lj78vv0l@developer.gserviceaccount.com",
+         "sub": "628645741881-noabiu23f5a8m8ovd8ucv698lj78vv0l@developer.gserviceaccount.com",
+         "aud": ["bookstore-esp-echo.cloudendpointsapis.com", "bookstore-esp-echo2.cloudendpointsapis.com"],
+         "iat": 1512754205,
+         "exp": 5112754205
+       }
+     )";
+
+TEST(AuthnUtilsTest, GetJwtPayloadFromHeaderTest) {
+  JwtPayload payload, expected_payload;
+  std::string value_base64 =
+      Base64::encode(kSecIstioAuthUserinfoHeaderValue.c_str(),
+                     kSecIstioAuthUserinfoHeaderValue.size());
+  Http::TestHeaderMapImpl request_headers_with_jwt{
+      {kSecIstioAuthUserInfoHeaderKey, value_base64}};
+  ASSERT_TRUE(Protobuf::TextFormat::ParseFromString(
+      R"(
+      user: "628645741881-noabiu23f5a8m8ovd8ucv698lj78vv0l@developer.gserviceaccount.com/628645741881-noabiu23f5a8m8ovd8ucv698lj78vv0l@developer.gserviceaccount.com"
+      audiences: ["bookstore-esp-echo.cloudendpointsapis.com"]
+      claims {
+        key: "aud"
+        value: "bookstore-esp-echo.cloudendpointsapis.com"
+      }
+      claims {
+        key: "iss"
+        value: "628645741881-noabiu23f5a8m8ovd8ucv698lj78vv0l@developer.gserviceaccount.com"
+      }
+      claims {
+        key: "sub"
+        value: "628645741881-noabiu23f5a8m8ovd8ucv698lj78vv0l@developer.gserviceaccount.com"
+      }
+    )",
+      &expected_payload));
+  // The payload returned from GetJWTPayloadFromHeaders() should be the same as
+  // the expected.
+  bool ret = AuthnUtils::GetJWTPayloadFromHeaders(
+      request_headers_with_jwt, LowerCaseString(kSecIstioAuthUserInfoHeaderKey),
+      &payload);
+  EXPECT_TRUE(ret);
+  EXPECT_TRUE(MessageDifferencer::Equals(expected_payload, payload));
+}
+
+TEST(AuthnUtilsTest, GetJwtPayloadFromHeaderWithNoAudTest) {
+  JwtPayload payload, expected_payload;
+  std::string value_base64 =
+      Base64::encode(kSecIstioAuthUserInfoHeaderWithNoAudValue.c_str(),
```

Comment:
Can we use this one? I though jwt need to use base64url encode/decode?

Reply:
There is no base64url encode function in istio/proxy repository; the existing code use base64 encode function to do the encoding. If there is no special chars in the input string that needed to be handled by base64url, base64 encode generates the same encoding result.

Reply:
Put a comment for that. Make these lines 100 - 104 into a function that return HeaderMap (or set the header map in the test fixture) so that it's easier to change the encode in the future if need to.

Reply:
Done.